src/libcore/ptr.rs

```diff
@@ -145,6 +148,31 @@ pub use intrinsics::write_bytes;
 /// // Ensure that the last item was dropped.
 /// assert!(weak.upgrade().is_none());
 /// ```
+///
+/// Drops a potentially unaligned value by copying it to aligned memory first:
+/// ```
+/// use std::ptr;
+/// use std::mem;
+///
+/// unsafe fn drop_after_copy<T>(to_drop: *mut T) {
+///     let mut copy: T = mem::uninitialized();
+///     let copy = &mut copy as *mut T;
+///     ptr::copy(to_drop, copy, 1);
+///     ptr::drop_in_place(copy);
+/// }
+///
+/// #[repr(packed, C)]
+/// struct Packed {
+///     _padding: u8,
+///     unaligned: Vec<i32>,
+/// }
+///
+/// let mut p = Packed { _padding: 0, unaligned: vec![42] };
+/// unsafe {
+///     drop_after_copy(&mut p.unaligned as *mut _);
+///     mem::forget(p);
+/// }
+/// ```
```

Comment:
This basically means that all `packed` structs with `Drop` fields have to be `Drop`, and that their `Drop` implementations might have to do this copy to aligned memory then drop "dance" :/

Reply:
AFAIK the compiler does this "dance" automatically for `packed` structs with `Drop` fields.

Reply:
Here's the code for that: https://github.com/rust-lang/rust/blob/0e98621e69890d67d906a436a68436d03a3edb89/src/librustc_mir/transform/add_moves_for_packed_drops.rs

Reply:
Wait, really? That would cause UB, though.

Reply:
Okay, yup, this is UB without `unsafe`, that's not good.

Reply:
Where do you see UB without `unsafe` here?

Reply:
@RalfJung https://github.com/rust-lang/rust/issues/54148